### PR TITLE
Add TF_AssignRefVariable

### PR DESCRIPTION
**Pull request by PatriceVignola**
The [Kernel Extension for Variable Operations API](https://github.com/tensorflow/community/blob/master/rfcs/20210504-kernel-extension-variable-ops.md) added pluggable device support for resource variables and incomplete support for ref variables (`TF_OpKernelContext_ForwardRefInputToRefOutput`), but it misses one endpoint to make ref variables usable: `TF_AssignRefVariable`.

This change attempts to fill the gaps in the RFC by adding `TF_AssignRefVariable`, which is analogous to `TF_AssignVariable` but for ref variables instead of resource variables. It uses the same semantics where the user has to pass a copy function to be called when copying the value tensor to the ref tensor.

### CODE REVIEW

File: tensorflow/c/kernels_experimental.cc [hunk 1]
```diff
@@ -194,6 +195,32 @@ void TF_AssignVariable(TF_OpKernelContext* ctx, int input_index,
   TF_SetStatus(status, TF_OK, "");
```
Comment (PatriceVignola) [inline tensorflow/c/kernels_experimental.cc:195]: I noticed that `TF_AssignVariable` unconditionally sets the status to `TF_OK`, so the way to check for errors is to check the status of the context. I'm not sure if it is an oversight, but this is the reason why I omitted this parameter in the ref version.

**Comment by wangpengmit** [comment]
I need more context. Please add a PR description and in it add a link to the original RFC(s) that introduced APIs such as `TF_AssignVariable`.

**Comment by PatriceVignola** [comment]
> I need more context. Please add a PR description and (in it) a link to the original RFC(s) that introduced APIs such as `TF_AssignVariable`.

@wangpengmit Got it, thanks! I updated the description.

File: tensorflow/c/kernels_experimental.cc (hunk 1 above)
Comment (wangpengmit) [reply tensorflow/c/kernels_experimental.cc:195]: It's probably a bug. The `copyFunc(ctx, tf_value, tf_tmp)` line should check `ctx->status()` after `copyFunc` returns. Please file a bug so we can fix `TF_AssignVariable` and `TF_AssignRefVariable` together later.

**Review by wangpengmit — APPROVED** [review]

File: tensorflow/c/kernels_experimental.cc (hunk 1 above)
Comment (PatriceVignola) [reply tensorflow/c/kernels_experimental.cc:195]: @wangpengmit Should I change the return type for `TF_AssignRefVariable` to `TF_Status` even if it's not used then, so that it's in line with `TF_AssignVariable`?

Comment (wangpengmit) [reply tensorflow/c/kernels_experimental.cc:195]: Oh yes please. I missed that part. We do want to change both to properly set return status later on.

File: tensorflow/c/kernels_experimental.h [hunk 2]
```diff
@@ -59,6 +59,12 @@ TF_CAPI_EXPORT extern void TF_AssignVariable(
                      TF_Tensor* dest),
     TF_Status* status);
 
+TF_CAPI_EXPORT extern void TF_AssignRefVariable(
```
Comment (wangpengmit) [inline tensorflow/c/kernels_experimental.h:62]: Please add a `TF_Status* status` output argument to `TF_AssignRefVariable`, in line with other API entries.

**Review by wangpengmit — CHANGES_REQUESTED** [review]

File: tensorflow/c/kernels_experimental.h (hunk 2 above)
Comment (PatriceVignola) [reply tensorflow/c/kernels_experimental.h:62]: Done!

**Review by wangpengmit — APPROVED** [review]

**Comment by PatriceVignola** [comment]
I'm not sure why the Windows Bazel GPU build is failing. I built it locally yesterday and it completed just fine, and the Linux GPU build completed successfully. Is it possible to look at the failures or is it unrelated to this PR?

**Comment by wangpengmit** [comment]
Hi @mihaimaruseac , is there a way to see the Windows Bazel GPU error messages?

**Comment by penpornk** [comment]
@wangpengmit Usually I just rerun the tests and hope the log appears. Will rerun them now since it's evening anyway.

**Comment by mihaimaruseac** [comment]
For Googlers, if you retrigger the job and click on the details link before the job finishes you are taken to the internal log page (external people will get an invalid link).

Once the job finishes, the link switched to the public version, but if the job fails in certain ways there is no public log. This is usually infra failure and could be ignored.

**Comment by PatriceVignola** [comment]
The build failed again without a "Details" link. Is there anything I can do to expedite this PR? I successfully built it and ran the tests locally but maybe I'm missing something. Is there a way to confirm whether it is an infra issue or not?

**Comment by penpornk** [comment]
@PatriceVignola I will manually import the PR and run the tests internally. I'll let you know if anything fails there.

**Comment by penpornk** [comment]
This PR was reverted because it broke internal tests. We will try to follow up with failure details soon.

**Comment by PatriceVignola** [comment]
Why was the PR merged in the first place? I thought you would only merge it if the internal tests passed. Please let us know what tests where failing when you have the info, I can debug it locally if I know what they are.

**Comment by penpornk** [comment]
@PatriceVignola We only run TensorFlow presubmit tests before merging. After merging, PRs could still be reverted if it broke TensorFlow nightly tests or other internal tests (e.g., unit / integration tests of internal applications that use TensorFlow). In this case, it broke an internal test.

**Comment by mihaimaruseac** [comment]
Here is the complete workflow for a PR:

7XNWZfbTauSmBXn

We have this many steps due to CI complexity and needing to support both multiple platforms x python versions x {GPU,CPU} and a requirement to have short presubmits to not negatively impact developer velocity.

**Comment by PatriceVignola** [comment]
Hi, I'm bumping this to know if it's possible to get the failure details, or the command line that was used to run the internal tests that failed. We are committed to investigate the issue and resolve it.

**Comment by penpornk** [comment]
Sorry for the delay. It took a while to extract a proxy test. Please add the following test case to [variable_ops_test.py](https://github.com/tensorflow/tensorflow/blob/d7307015d3ab7ad82305c46c259fc711c3948ce8/tensorflow/python/kernel_tests/variables/variable_ops_test.py#L65)

```python
  @test_util.run_deprecated_v1
  def testString(self):
    data = array_ops.stack([b"data"])
    buffer_var = variables.VariableV1(
        initial_value=array_ops.zeros(shape=(), dtype=dtypes.string),
        trainable=False,
        collections=[ops.GraphKeys.LOCAL_VARIABLES],
        name="buffer",
        dtype=dtypes.string,
        validate_shape=False,
        use_resource=False)
    result = state_ops.assign(buffer_var, data, validate_shape=False)
    with self.cached_session() as sess:
      sess.run(variables.local_variables_initializer())
      self.assertEqual(result.eval(), b"data")
```
To run:
```
bazel test --config=opt //tensorflow/python/kernel_tests/variables:variable_ops_test --test_filter=testString
```

**Comment by penpornk** [comment]
Oops sorry. I forgot to post the error message. I'm not sure why it's happening either. I don't see anything apparently wrong in your refactor code.
```
tensorflow/core/framework/tensor.cc:725] Check failed: dtype() == expected_dtype (1 vs. 7) string expected, got float
```

I could reproduce the error with the TensorFlow docker image [here](https://www.tensorflow.org/install/source#cpu-only).
The error happened with commit da9a9d69bb1c98f0e88883b3d0274b139d455e49 (merged pull request) but not the commit before (2d040ce3dde8553125eaaa49a97355cc2e05a445).

To summarize, these series of command should be able to reproduce the error:
```
$ docker pull tensorflow/tensorflow:devel
$ docker run -it -w /tensorflow_src -v $PWD:/mnt -e HOST_PERMS="$(id -u):$(id -g)" \
    tensorflow/tensorflow:devel bash
$ git pull
$ git checkout da9a9d69bb1c98f0e88883b3d0274b139d455e49
$ yes "" | ./configure
$ bazel test --config=opt //tensorflow/python/kernel_tests/variables:variable_ops_test --test_filter=testString
```

**Comment by PatriceVignola** [comment]
I fixed the problem: https://github.com/tensorflow/tensorflow/pull/55640

It was a silly mistake of using `DCHECK` instead of `CHECK` on like 77 in ref_var.cc. No idea why I changed the macro, I think I may have accidentally deleted the D while removed leading spaces.

**Comment by mihaimaruseac** [comment]
`CHECK`s have been causing issues in the past and are discouraged (there is automation that blocks submit with them, unless there are additional steps taken). `DCHECK` is a workaround that some people use without fully acknowledging all the differences between them.

`OP_REQUIRES` is the best approach to use.

**Comment by PatriceVignola** [comment]
It would have been fine if the submission had been blocked, but the automation silently changing `CHECK` to `DCHECK` can be hard to debug when the change gets reverted, especially when refactoring old code that already contained them.

**Comment by mihaimaruseac** [comment]
It wasn't automation, it was a deliberate change during PR import.

**Comment by penpornk** [comment]
Sorry for the late reply and thank you for looking into it! 
Yes, it was me changing `CHECK` to `DCHECK` during merge. Sorry I forgot to mention that. :(
